algorithms/src/std_algorithms/impl/Kokkos_RandomAccessIterator.hpp

```diff
@@ -111,77 +118,87 @@ class RandomAccessIterator< ::Kokkos::View<DataType, Args...> > {
+      m_data += n;
+    else
+      m_data += n * m_stride;
     return *this;
   }
 
   KOKKOS_FUNCTION
   iterator_type& operator-=(difference_type n) {
-    m_current_index -= n;
+    if constexpr (is_always_contiguous)
+      m_data -= n;
+    else
+      m_data -= n * m_stride;
     return *this;
   }
 
   KOKKOS_FUNCTION
   iterator_type operator+(difference_type n) const {
-    return iterator_type(m_view, m_current_index + n);
+    auto it = *this;
+    it += n;
+    return it;
   }
 
   KOKKOS_FUNCTION
   iterator_type operator-(difference_type n) const {
-    return iterator_type(m_view, m_current_index - n);
+    auto it = *this;
+    it -= n;
+    return it;
   }
 
   KOKKOS_FUNCTION
   difference_type operator-(iterator_type it) const {
-    return m_current_index - it.m_current_index;
+    if constexpr (is_always_contiguous)
+      return m_data - it.m_data;
+    else
+      return (m_data - it.m_data) / m_stride;
   }
 
   KOKKOS_FUNCTION
-  bool operator==(iterator_type other) const {
-    return m_current_index == other.m_current_index &&
-           m_view.data() == other.m_view.data();
-  }
+  bool operator==(iterator_type other) const { return m_data == other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator!=(iterator_type other) const {
-    return m_current_index != other.m_current_index ||
-           m_view.data() != other.m_view.data();
-  }
+  bool operator!=(iterator_type other) const { return m_data != other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator<(iterator_type other) const {
-    return m_current_index < other.m_current_index;
-  }
+  bool operator<(iterator_type other) const { return m_data < other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator<=(iterator_type other) const {
-    return m_current_index <= other.m_current_index;
-  }
+  bool operator<=(iterator_type other) const { return m_data <= other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator>(iterator_type other) const {
-    return m_current_index > other.m_current_index;
-  }
+  bool operator>(iterator_type other) const { return m_data > other.m_data; }
 
   KOKKOS_FUNCTION
-  bool operator>=(iterator_type other) const {
-    return m_current_index >= other.m_current_index;
-  }
+  bool operator>=(iterator_type other) const { return m_data >= other.m_data; }
+
+  KOKKOS_FUNCTION
+  reference operator*() const { return *m_data; }
 
   KOKKOS_FUNCTION
-  reference operator*() const { return m_view(m_current_index); }
+  pointer data() const { return m_data; }
 
   KOKKOS_FUNCTION
-  view_type view() const { return m_view; }
+  int stride() const { return m_stride; }
 
  private:
-  view_type m_view;
-  ptrdiff_t m_current_index = 0;
+  pointer m_data;
+  int m_stride;
+  static constexpr bool is_always_contiguous =
+      (std::is_same_v<typename view_type::traits::array_layout,
+                      Kokkos::LayoutLeft> ||
+       std::is_same_v<typename view_type::traits::array_layout,
+                      Kokkos::LayoutRight>);
```

Comment:
I'm curious if this check really improves performance visibly.

Reply:
See tables in the description.